cumulus/parachains/common/src/xcm_config.rs

```diff
@@ -78,3 +85,137 @@ impl<Location: Get<MultiLocation>> ContainsPair<MultiAsset, MultiLocation>
 		matches!(asset.id, Concrete(ref id) if id == origin && origin == &Location::get())
 	}
 }
+
+/// Allows execution from `origin` (only with `OriginKind::Xcm`) if it is contained in
+/// `AllowedOrigin` and if it is just a straight `Transact` which passes `AllowedCall` matcher.
+pub struct AllowUnpaidTransactsFrom<RuntimeCall, AllowedCall, AllowedOrigin>(
+	sp_std::marker::PhantomData<(RuntimeCall, AllowedCall, AllowedOrigin)>,
+);
+impl<
+		RuntimeCall: Decode,
+		AllowedCall: Contains<RuntimeCall>,
+		AllowedOrigin: Contains<MultiLocation>,
+	> ShouldExecute for AllowUnpaidTransactsFrom<RuntimeCall, AllowedCall, AllowedOrigin>
+{
+	fn should_execute<Call>(
+		origin: &MultiLocation,
+		instructions: &mut [Instruction<Call>],
+		_max_weight: Weight,
+		_properties: &mut xcm_executor::traits::Properties,
+	) -> Result<(), ProcessMessageError> {
+		log::trace!(
+			target: "xcm::barriers",
+			"AllowUnpaidTransactFrom origin: {:?}, instructions: {:?}, max_weight: {:?}, properties: {:?}",
+			origin, instructions, _max_weight, _properties,
+		);
+
+		// we only allow from configured origins
+		ensure!(AllowedOrigin::contains(origin), ProcessMessageError::Unsupported);
+
+		// we expect an XCM program with single `Transact` call
+		instructions
+			.matcher()
+			.assert_remaining_insts(1)?
+			.match_next_inst(|inst| match inst {
+				Transact { origin_kind: OriginKind::Xcm, call: encoded_call, .. } => {
```

Comment:
XCM v3 and greater require an explicit `UnpaidExecution` instruction to allow unpaid execution. Matching on a program with just a single instruction that is `Transact` seems to break this pattern. I would still enforce the same pattern, especially as this is in `common`, it can be picked up by any system parachain. If this pattern is necessary and specific to bridges, it should not live in `common` and be documented as such. @franciscoaguirre